content/sealed-types.md

```diff
@@ -0,0 +1,278 @@
+---
+layout: sip
+title: SIP-NN - Sealed Types
+vote-status: pending
+permalink: /sips/:title.html
+redirect_from: /sips/pending/sealed-types.html
+---
+
+**By: Dale Wijnand and Fengyun Liu**
+
+## History
+
+| Date          | Version   |
+|---------------|-----------|
+| May 18th 2021 | Submitted |
+
+## Introduction
+
+Exhaustivity checking is one of the safety belts for using pattern matching in functional
+programming.  However, if one wants to partition values of an existing type to reason about them in
+separate cases of a match or as separate types this requires wrapping the values in new classes,
+which incurs a boxing cost.
+
+As an alternative, one may define custom extractors and use them as the case patterns. However, there is no
+way in Scala to declare that these extractors are complementary, i.e. the match is exhaustive when
+the complete set of extractors are used together.
+
+Similarly, one can define `TypeTest`s for matching abstract types, but there is no way to determine
+if a match on an abstract type exhausts all its subtypes.
+
+This SIP solves these three problems by introducing *sealed types*.
+
+## Motivating Examples
+
+We've identified several real world use cases that calls for enhancing exhaustivity checking, and
+used them to stress the design proposed here.  You can find them [here][problems], but we'll present
+an example below.
+
+Using the opaque types and custom extractors we can work safely with positive integers and negative
+integers:
+
+```scala
+opaque type Pos <: Int = Int
+opaque type Neg <: Int = Int
+
+object Pos { def unapply(x: Int): Option[Pos] = if (x > 0) Some(x) else None }
+object Neg { def unapply(x: Int): Option[Neg] = if (x < 0) Some(x) else None }
+
+(n: Int) match
+  case 0      =>
+  case Pos(x) =>
+  case Neg(x) =>
+```
+
+With the above, when we get a `Pos` value, it's guaranteed to be a positive number. The same goes
+for `Neg`. Sadly the match is reported as not exhaustive because the two extractors and the value
+`0` aren't known to be complementary.
+
+## Design
+
+We identify two root causes:
+
+1. You can't define a type, that isn't a class, as `sealed`
+2. You can't define a mapping from values to types
+
+The *sealed type*, proposed by this SIP, allow partitioning of value of a given type into a sealed
+type hierarchy.  [Here][solutions] you can find how sealed types address the issues faced by all the
+motivating examples, but we'll present here how it fixes the number example above.
+
+In order to partition int into positive, zero, and negative numbers we'll define a new `sealed` type
+`Num` and how to distinguish its subtypes with match syntax:
+
+```scala
+sealed type Num = Int {
+  case 0          => val Zero
+  case n if n > 0 => type Pos
+  case _          => type Neg
+}
```

Comment:
I suspect this syntax is very confusing. I run a very simple test. I showed the snippet above to 6 folks who have experience both in Scala 2 and Scala 3 but do not know about this proposal.
None of them could guess what the snippet is supposed to do. After revealing what it desugars to, they unanimously agreed that it would be a constant source of confusion.

Firstly, I believe that even years after it landing in the language, googling "sealed types scala" would return results that are overwhelmingly about sealed traits. Sealed traits are in language much longer, there is much more written about them, all their descriptions probably contain the word "types" somewhere, and search engines are not very good at interpreting subtle differences in queries.

Another aspect here is that the syntax looks like a mashup between refined types, pattern matching and match types. this will definitely add to the confusion.

Lastly, and this is not a fatal flaw, but `case 0 => val Zero` looks really weird and, in my opinion, not very scala-like. I think it can be substituted with `case 0 => 0.type` or even just `case 0 => type Zero` it would make much more sense.